kong/plugins/hmac-auth/access.lua

```diff
@@ -0,0 +1,134 @@
+local cache = require "kong.tools.database_cache"
+local stringy = require "stringy"
+local responses = require "kong.tools.responses"
+local constants = require "kong.constants"
+
+local math_abs = math.abs
+local ngx_time = ngx.time
+local ngx_gmatch = ngx.re.gmatch
+local ngx_decode_base64 = ngx.decode_base64
+local ngx_parse_time = ngx.parse_http_time
+local ngx_sha1 = ngx.hmac_sha1
+local ngx_set_header = ngx.req.set_header
+
+local AUTHORIZATION = "authorization"
+local PROXY_AUTHORIZATION = "proxy-authorization"
+local DATE = "date"
+local SIGNATURE_NOT_VALID = "HMAC signature cannot be verified"
+
+local _M = {}
+
+local function retrieve_hmac_fields(request, header_name, conf)
+  local username, signature, algorithm
+  local authorization_header = request.get_headers()[header_name]
+  if authorization_header then
+    -- Authentication: hmac username:base64(hmac-sha1(Date))
+    local iterator, iter_err = ngx_gmatch(authorization_header, "\\s*[Hh]mac\\s*(.+)")
+    if not iterator then
+      ngx.log(ngx.ERR, iter_err)
+      return
+    end
+
+    local m, err = iterator()
+    if err then
+      ngx.log(ngx.ERR, err)
+      return 
+    end
+    
+    if m and table.getn(m) > 0 then
+      local hmac_fields = stringy.split(m[1], ":")
+      if hmac_fields and #hmac_fields > 2 then
+        username = hmac_fields[1]
+        signature = ngx_decode_base64(hmac_fields[2])
+        algorithm = hmac_fields[3]
+      end  
+    end
+  end
+    
+  if conf.hide_credentials then
+    request.clear_header(header_name)
+  end
+  
+  return username, signature, algorithm
+end
+
+local function validate_signature(request, secret, signature, algorithm, defaultClockSkew)
```

Comment:
@harlow  for now only sha1 would be supported and would be mentioned in the manual, here argument is optional field and any value passed by user will be discarded.